### PR TITLE
feat(disc): add DnsDiscovery service

### LINKED ISSUE #779: Add dns discovery service
### Describe the feature

Add https://eips.ethereum.org/EIPS/eip-1459#dns-record-structure DNS discovery

Should be a service task, same as discv4 that emits `NodeRecord`s

ref #768

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
Closes #779

implements Dns Discovery that
* discovers trees
* emits `NodeRecord`
* rate limited 3 req/s (same geth)

Also adds `Chain::public_dns_network`

See `test_dns_resolver` that discovers via public DNS network

Will integrate this with network in followup.

### TODO
- [x] tests

### CODE REVIEW

**Review by onbjerg — COMMENTED** [review]
I think it would be cool if we could specify multiple domains to crawl. For the integration with network we should also consider how we can add separate discv4 and dnsdisc metrics so we can track where peers are coming from etc.

Edit: I see the first one is already possible, very nice

File: crates/net/dns/src/lib.rs [hunk 1]
```diff
@@ -73,38 +146,368 @@ impl DnsDiscoveryService {
...
+            Ok((root, link)) => match self.trees.entry(link.clone()) {
+                Entry::Occupied(mut entry) => {
+                    entry.get_mut().update_root(root);
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(SyncTree::new(root, link));
+                }
+            },
+            Err((err, link)) => {
+                debug!(?err, ?link, "Failed to lookup root")
```
Comment (onbjerg) [inline crates/net/dns/src/lib.rs:202]: nit: add a log target

**Review by onbjerg — APPROVED** [review]
LGTM overall, just a small nit. Excited for this

File: crates/net/dns/src/lib.rs [hunk 2]
```diff
@@ -73,38 +146,368 @@ impl DnsDiscoveryService {
...
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Poll::Ready(Some(ready!(self.get_mut().poll(cx))))
+    }
+}
+
+/// Commands sent from [DnsDiscoveryHandle] to [DnsDiscoveryService]
+enum DnsDiscoveryCommand {
+    /// Sync a tree
+    SyncTee(LinkEntry),
```
Comment (Rjected) [inline crates/net/dns/src/lib.rs:335]: nit: `SyncTree`

**Review by Rjected — APPROVED** [review]
smol typo otherwise lgtm! very cool

**Pull request merged**